src/main/kotlin/org/rust/ide/refactoring/introduceConstant/RsIntroduceConstantHandler.kt

```diff
@@ -0,0 +1,99 @@
+import com.intellij.refactoring.RefactoringBundle
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import org.rust.ide.inspections.import.AutoImportFix
+import org.rust.ide.refactoring.*
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.types.type
+import org.rust.openapiext.runWriteCommandAction
+
+class RsIntroduceConstantHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext) {
+        if (file !is RsFile) return
+        val exprs = findCandidateExpressionsToExtract(editor, file).filter { it.isExtractable() }
+
+        when (exprs.size) {
+            0 -> {
+                val message = RefactoringBundle.message(if (editor.selectionModel.hasSelection())
+                    "selected.block.should.represent.an.expression"
+                else
+                    "refactoring.introduce.selection.error"
+                )
+                val title = RefactoringBundle.message("introduce.variable.title")
+                val helpId = "refactoring.extractVariable"
+                CommonRefactoringUtil.showErrorHint(project, editor, message, title, helpId)
+            }
+            1 -> extractExpression(editor, exprs.single())
+            else -> {
+                showExpressionChooser(editor, exprs) {
+                    extractExpression(editor, it)
+                }
+            }
+        }
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        //this doesn't get called form the editor.
+    }
+}
+
+private fun RsExpr.isExtractable(): Boolean {
+    return when (this) {
+        is RsLitExpr -> true
+        is RsBinaryExpr -> this.left.isExtractable() && (this.right?.isExtractable() ?: true)
+        else -> false
+    }
+}
+
+private fun replaceWithConstant(expr: RsExpr, occurrences: List<RsExpr>, candidate: InsertionCandidate, editor: Editor) {
+    val project = expr.project
+    val factory = RsPsiFactory(project)
+    val suggestedNames = expr.suggestedNames()
+    val name = suggestedNames.default.toUpperCase()
+    val const = factory.createConstant(name, expr)
+
+    val (inserted, replaced) = project.runWriteCommandAction {
+        val newline = PsiParserFacade.SERVICE.getInstance(project).createWhiteSpaceFromText("\n")
+        val context = candidate.parent
+        val inserted = context.addBefore(const, candidate.anchor)
+        context.addAfter(newline, inserted)
+        val replaced = occurrences.map {
+            val created = factory.createExpression(name)
+            val element = it.replace(created) as RsElement
+            AutoImportFix(element, AutoImportFix.Type.GENERAL_PATH).invoke(project)
+            element
+        }
+        Pair(inserted as RsConstant, replaced.toList())
+    }
+
+    editor.caretModel.moveToOffset(inserted.identifier?.textRange?.startOffset ?: 0)
```

Comment:
Looks like ` ?: 0` is impossible, but it can leads to fun bugs if it will become possible somehow in the future. Maybe it's better to fail explicitly `?: error("Impossible because we just created a constant with a name")`